src/threading/cthread.cpp

```diff
@@ -132,54 +133,53 @@ void task_pool::wait(task_group* g) {
 }
 
 [[noreturn]]
-static void terminate(const char *const msg) {
+static void terminate(std::string msg) {
     std::cerr << "NMC_NUM_THREADS_ERROR: " << msg << std::endl;
     std::terminate();
 }
 
 // should check string, throw exception on missing or badly formed
 static size_t global_get_num_threads() {
-    const char* nthreads_str;
+    const char* str;
+
     // select variable to use:
     //   If NMC_NUM_THREADS_VAR is set, use $NMC_NUM_THREADS_VAR
     //   else if NMC_NUM_THREAD set, use it
     //   else if OMP_NUM_THREADS set, use it
     if (auto nthreads_var_name = std::getenv("NMC_NUM_THREADS_VAR")) {
-        nthreads_str = std::getenv(nthreads_var_name);
+        str = std::getenv(nthreads_var_name);
     }
-    else if (! (nthreads_str = std::getenv("NMC_NUM_THREADS"))) {
-        nthreads_str = std::getenv("OMP_NUM_THREADS");
+    else if (! (str = std::getenv("NMC_NUM_THREADS"))) {
+        str = std::getenv("OMP_NUM_THREADS");
     }
 
     // If the selected var is unset,
     //   or no var is set,
     //   error
-    if (! nthreads_str) {
-        terminate("No environmental var defined");
-     }
-
-    // only composed of spaces*digits*space*
-    auto nthreads_str_end{nthreads_str};
-    while (std::isspace(*nthreads_str_end)) {
-        ++nthreads_str_end;
-    }
-    while (std::isdigit(*nthreads_str_end)) {
-        ++nthreads_str_end;
-    }
-    while (std::isspace(*nthreads_str_end)) {
-        ++nthreads_str_end;
-    }
-    if (*nthreads_str_end) {
-        terminate("Num threads is not a single integer");
+    if (!str) {
+        std::cerr << "WARNING: Environment variable describing number of threads not set: using 1 thread.\n";
+        std::cerr << "Set one of NMC_NUM_THREADS_VAR, NMC_NUM_THREADS or OMP_NUM_THREADS to set the\n";
+        std::cerr << "number of threads.\n";
+        return 1;
     }
 
-    // and it's got a single non-zero value
-    auto nthreads{std::atoi(nthreads_str)};
-    if (! nthreads) {
-        terminate("Num threads is not a non-zero number");
+    try {
+        // The following with throw if:
+        // * the call to stoul() can't perform string->unsigned conversion;
+        // * the parsed number is followed by any non-whitespace characters.
+        auto end = std::size_t();
+        auto nthreads = std::stoul(str, &end);
+        while (end<std::strlen(str)) {
+            if (!std::isspace(str[end])) {
+                throw int();
+            }
+            ++end;
+        }
+        return nthreads;
+    }
+    catch(...) {
```

Comment:
`std::stoi` throws an exception if it can't perform the conversion to unsigned int.
This is just a catch all for "something went wrong in parsing value".